builder/haddock-builder.nix

```diff
@@ -0,0 +1,180 @@
+, preHaddock
+, postHaddock
+, setupInstallFlags
+
+, doHoogle
+, hyperlinkSource
+, setupHaddockFlags
+, setupGhcOptions
+
+, needsProfiling
+, componentDrv
+, configFiles # component config files
+}:
+
+let
+  # the target dir for haddock documentation
+  docdir = docoutput: docoutput + "/share/doc/" + componentId.cname;
+
+  packageCfgDir = configFiles.packageCfgDir;
+
+  fullName = if haskellLib.isAll componentId
+    then "${name}-docs-all"
+    else "${name}-${componentId.ctype}-${componentId.cname}-docs";
+
+  docsConfigFiles = makeConfigFiles {
+    inherit (package) identifier;
+    inherit component fullName flags needsProfiling;
+    chooseDrv = p: p.haddock;
+  };
+
+  # let
+  #   # depDocConfig = map (x: x.configFiles) (haskellLib.flatLibDepends component);
+  # in runCommand "${ghc.targetPrefix}${fullName}-docs-config" {
+  #   nativeBuildInputs = [ghc];
+  # } (''
+  #   mkdir -p $out
+
+  #   for i in cabal.config configure-flags ghc-environment; do
+  #     substitute "${configFiles}/$i" "$out/$i"  --replace "${configFiles}" "$out"
+  #   done
+
+  #   mkdir -p $out/${packageCfgDir}
+
+  #   # Copy over the nonReinstallablePkgs from the global package db.
+  #   ${lib.concatMapStringsSep "\n" (p: ''
+  #     find ${ghc}/lib/${ghc.name}/package.conf.d -name '${p}*.conf' -exec cp -f {} $out/${packageCfgDir} \;
+  #   '') nonReinstallablePkgs}
+
+  # '');
+
+  finalConfigureFlags = lib.concatStringsSep " " (
+    [ "--prefix=${componentDrv}"
+      "--docdir=${docdir "$doc"}"
+      "${haskellLib.componentTarget componentId}"
+      "$(cat ${docsConfigFiles}/configure-flags)"
+      # GHC
+      "--with-ghc=${ghc.targetPrefix}ghc"
+      "--with-ghc-pkg=${ghc.targetPrefix}ghc-pkg"
+      "--with-hsc2hs=${ghc.targetPrefix}hsc2hs"
+    ]
+      ++ configureFlags
+      ++ (ghc.extraConfigureFlags or [])
+    );
+
+  shellWrappers = ghcForComponent {
+    componentName = fullName;
+    configFiles = docsConfigFiles;
+  };
+
+in stdenv.lib.fix (drv: stdenv.mkDerivation ({
+  name = "${ghc.targetPrefix}${fullName}";
+  inherit src;
+
+  LANG = "en_US.UTF-8";         # GHC needs the locale configured during the Haddock phase.
+  LC_ALL = "en_US.UTF-8";
+
+  passthru = {
+    configFiles = docsConfigFiles;
+
+    # The directory containing the haddock documentation.
+    # `null' if no haddock documentation was built.
+    haddockDir = "${docdir drv.doc}/html";
+  };
+
+  outputs = ["out" "doc"];
+
+  enableParallelBuilding = true;
+
+  nativeBuildInputs =
+    [ shellWrappers buildPackages.removeReferencesTo ]
+    ++ componentDrv.executableToolDepends;
+
+  SETUP_HS = setup + /bin/Setup;
+
+  configurePhase = ''
+    echo Configure flags:
+    printf "%q " ${finalConfigureFlags}
+    echo
+    $SETUP_HS configure ${finalConfigureFlags}
+  '';
+
+  buildPhase = ''
+    runHook preHaddock
+    # If we don't have any source files, no need to run haddock
+    [[ -n $(find . -name "*.hs" -o -name "*.lhs") ]] && {
+
+    $SETUP_HS haddock \
```

Comment:
This will build the component in question, right? So this will potentially end up with us building every component twice.

Reply:
And this time without e.g. the `-j` flags we pass to `build` in the main builder.

So I wonder: what if we just used the main component builder, but ran it twice, once with Haddock enabled and once without?
- Code is all shared, so we don't need to duplicate all this stuff.
- We build everything twice, but I think we do that in this approach too.
- Easy to provide a module option to just go back to the combined build if the cost of building everything twice is too painful.

Reply:
> This will build the component in question, right?

No. `setup haddock` only build dependencies (not the package itself) and @purefn has set it up so that it reuses the non haddock derivations for its dependencies.

I think the only overhead that is duplicated is the work done by `setup configure`!

Reply:
Actually I suspect only `cabal haddock` would build dependencies.

Reply:
Huh, you're right. I was misled by the output and it does indeed not rebuild it! Phew.

I do wonder whether we could cut down the code duplication by reusing the component builder with `dontBuild=true` or something?

Reply:
I'm quite keen to keep the haddock stuff in a different file to keep the size of comp-builder.nix down.  Also modifying haddock-builder.nix should never result in long rebuilds, where as ghc 8.10 depends on ghc 8.8.3 depends on alex depends on comp-builder.nix, so it will be nice to be able to modify the haddock derivation without fear of long rebuilds.

I have tried to reduce the amount of duplication though so I think there is a lot less duplication of the `comp-builder.nix` code in `haddock-builder.nix` now.